### DISCUSSION TITLE
Optimizing useController performance

### DISCUSSION THREAD

**Discussion by orenklein** [react-hook-form/react-hook-form#5569] [Q&A]
As described in the documentation, `FormProvider` has performance cost due to usage of Context API.
A component that use `useController` will rerender every time something is changed in the form (changes in other inputs).
In most cases the component that use `useController` renders some UI framework (Material UI, AntD, etc..) input component.

The goal is to at least memoize the input component props so it will rerender only for self subscribed changes.

First step would be to wrap the input component with `React.memo`. 
Unfortunately, `useController` returns new `field` props every rerender - `ref`, `onBlur`, `onChange`, `value`. (Note: Those `field` props used to be memoized with `useCallback` in v6).

I wrote a `useMemoizedController` which worked until version `7.8.0`:
```typescript
export const useMemoizedController= <TFieldValues extends FieldValues = FieldValues>(
  options: UseControllerProps<TFieldValues>
): UseControllerReturn<TFieldValues> => {
  const { field, ...rest } = useController({ ...options });
  const onBlur = React.useMemo(() => field.onBlur, [options.name]);
  const onChange = React.useMemo(() => field.onChange, [options.name]);
  const ref = React.useMemo(() => field.ref, [options.name]);
  return { field: { ...field, onBlur, onChange, ref }, ...rest };
};
```

In version `7.8.0` it stopped working successfully, it seems that `ref` is changed in each render.
You can see in https://codesandbox.io/s/infallible-river-v8thp?file=/src/App.tsx:
1. Click on First Name select input and choose Name1
2. Focus on something else on the screen like the second input line (to trigger onBlur)
3. required error is seen even though the field is updated

Other steps to further investigate the behavior:
1. uncomment line 33 to replace the current behavior, and repeat on the above steps - it will be normal.
2. uncomment the next lines (with the console.log inside), and repeat on the above steps, it will log the instance with the initial value (empty string). (same will be if sending defaultValue to ControlledSelect, with defaultValue as the initial value) 

I assume it happened in this commit https://github.com/react-hook-form/react-hook-form/commit/14db607ab7476bbc7fd8297752b01e230a8552b4#diff-3f5bdfa7d822b4562b8e1c72dd1996197fe93478b039779bb78853ab3ae2a3fe

Do you have any suggestion to achieve the needed behavior - memoized props for the inner input component?

**Comment by bluebill1049** [comment]
relevant issue: https://github.com/react-hook-form/react-hook-form/issues/5591

**Comment by orenklein** [comment]
My workaround for now is adding `field.value` to ref's useMemo dependencies:
```typescript 
const ref = React.useMemo(() => field.ref, [options.name, field.value]);
